source/extensions/filters/http/cache/cache_policy.h

```diff
@@ -0,0 +1,98 @@
+ */
+class CachePolicy {
+public:
+  virtual ~CachePolicy() = default;
+
+  /**
+   * Calculates the lookup key for storing the entry in the cache.
+   * @param request_headers - headers from the request the CacheFilter is currently processing.
+   */
+  virtual Key createCacheKey(const Http::RequestHeaderMap& request_headers) PURE;
+
+  /**
+   * Modifies the cacheability of the response during decoding.
+   * @param request_headers - headers from the request the CacheFilter is currently processing.
+   * @param request_cache_control - the result of parsing the request's Cache-Control header, parsed
+   * by the caller.
+   * @return true if the response may be cached, based on the contents of the request.
+   */
+  virtual bool requestCacheable(const Http::RequestHeaderMap& request_headers,
+                                const RequestCacheControl& request_cache_control) PURE;
+
+  /**
+   * Modifies the cacheability of the response during encoding.
+   * @param request_headers - headers from the request the CacheFilter is currently processing.
+   * @param response_headers - headers from the upstream response the CacheFilter is currently
+   * processing.
+   * @param response_cache_control - the result of parsing the response's Cache-Control header,
+   * parsed by the caller.
+   * @param vary_allow_list - list of headers that the cache will respect when creating the Key for
+   * Vary-differentiated responses.
+   * @return true if the response may be cached.
+   */
+  virtual bool responseCacheable(const Http::RequestHeaderMap& request_headers,
+                                 const Http::ResponseHeaderMap& response_headers,
+                                 const ResponseCacheControl& response_cache_control,
+                                 const VaryHeader& vary_allow_list) PURE;
+
+  /**
+   * Calculates whether the cached entry may be used directly or must be validated with upstream.
+   * @param request_headers - headers from the request the CacheFilter is currently processing.
+   * @param response_headers - headers from the cached response the CacheFilter has retrieved.
+   * @param request_cache_control - the parsed result of the request's Cache-Control header, parsed
+   * by the caller.
+   * @param cached_response_cache_control - the parsed result of the response's Cache-Control
+   * header, parsed by the caller.
+   * @param content_length - the byte length of the cached content.
+   * @param cached_metadata - the metadata that has been stored along side the cached entry.
+   * @param now - the timestamp for this request.
+   * @return details about whether or not the cached entry can be used.
+   */
+  virtual CacheEntryUsability
+  computeCacheEntryUsability(const Http::RequestHeaderMap& request_headers,
+                             const Http::ResponseHeaderMap& cached_response_headers,
+                             const RequestCacheControl& request_cache_control,
```

Comment:
dumb question: why do we need to have both the request/response headers and parsed request/response cache-control headers? Isn't cache-control header already in the request/response headers?

Reply:
I was thinking the interface shouldn't make assumptions for what a cache policy can use to determine cacheability. For example, a custom header `X-Company-Customheader` might contain information the CachePolicy wants to incorporate into the cacheability evaluation.